caiman/source_extraction/cnmf/lazy_arrays.py

```diff
@@ -0,0 +1,439 @@
+from warnings import warn
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import *
+from itertools import product as iter_product
+from time import time
+
+
+import numpy as np
+from scipy.sparse import csc_matrix
+import h5py
+
+
+# TODO: Is there a better way to do this?
+slice_or_int_or_range = Union[int, slice, range]
+
+
+class LazyArray(ABC):
+    """
+    Base class for arrays that exhibit lazy computation upon indexing
+    """
+    @property
+    def dtype(self) -> np.dtype:
+        """
+        np.dtype: numpy dtype
+        """
+        # compute the slice at the first index and return the dtype of that
+        # TODO: there might be a nicer way to do this
+        return self[0].dtype
+
+    @property
+    @abstractmethod
+    def shape(self) -> Tuple[int, int, int]:
+        """
+        Tuple[int]: (n_frames, dims_x, dims_y)
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def n_frames(self) -> int:
+        """
+        int: number of frames
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def min(self) -> float:
+        """
+        float: min value of the array if it were fully computed
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def max(self) -> float:
+        """
+        float: max value of the array if it were fully computed
+        """
+        pass
+
+    @property
+    def ndim(self) -> int:
+        """
+        int: Number of dimensions
+        """
+        return len(self.shape)
+
+    @property
+    def nbytes(self) -> int:
+        """
+        int: number of bytes required for the array if it were fully computed
+        """
+        return np.prod(self.shape + (np.dtype(self.dtype).itemsize,))
+
+    @property
+    def nbytes_gb(self) -> float:
+        """
+        float: number of gigabytes required for the array if it were fully computed
+        """
+        return self.nbytes / 1e9
+
+    @abstractmethod
+    def _compute_at_indices(self, indices: Union[int, slice]) -> np.ndarray:
+        """
+        Lazy computation logic goes here. Computes the array at the desired indices.
```

Comment:
Is this leaving room for future work?

Reply:
this is just an abstract method indicating that it must be implemented in the subclass to handle the lazy compute logic for the given slice indices.